lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -660,6 +644,87 @@ bool mlir::triton::preProcessLoopAndGetSchedule(
+  auto countInsertsBetween = [](Operation *op1, Operation *op2) {
+    int count = 0;
+    for (auto op = op1; op != op2; op = op->getNextNode()) {
+      if (isa<ttg::InsertSliceAsyncOp>(op))
+        count++;
+      // Intentionally skip block ops' children. This will give us
+      // convervatively low number of insert ops.
+    }
+    return count;
+  };
+
+  int minWaitNumber = INT_MAX;
+
+  // DFS the def chain of the extract op to find the insert op. On each path
+  // we calculate the number of insert ops. Then we select the minimum number
+  // of insert ops among all the paths.
+  std::function<int(Value, Operation *, int)> minOverHistories =
+      [&](Value val, Operation *sinkOp, int thisHistorySum) -> int {
+    if (auto defOp = val.getDefiningOp()) {
+      if (isa<ttg::InsertSliceAsyncOp, ttng::InsertSliceTMAOp>(defOp)) {
+        thisHistorySum += countInsertsBetween(defOp->getNextNode(), sinkOp);
+        minWaitNumber = std::min(minWaitNumber, thisHistorySum);
+        return minWaitNumber;
+      }
+      // We cannot track further. Conservatively return 0, and assert in debug,
+      // so that we can easier find unsupported cases.
+      assert(false && "Found source op that is not InsertSlice.");
+      return 0;
+    }
+    if (auto arg = val.dyn_cast<BlockArgument>()) {
+      auto block = arg.getOwner();
+      auto currForOp = dyn_cast<scf::ForOp>(block->getParentOp());
+
+      // We cannot track further. Conservatively return 0, and assert in debug,
+      // so that we can easier find unsupported cases.
+      assert(currForOp == forOp && "Unexpected forOp in minOverHistories.");
+      if (currForOp != forOp)
+        return 0;
+
+      auto firstForInst = &*currForOp.getBody()->begin();
+      auto insertsBetween = countInsertsBetween(firstForInst, sinkOp);
+      thisHistorySum += insertsBetween;
+      if (thisHistorySum >= minWaitNumber)
+        return minWaitNumber;
+
+      // get the value value assigned to the argument coming from outside the
+      // loop
+      auto incomingVal = currForOp.getInitArgs()[arg.getArgNumber() - 1];
+      int min1 = minOverHistories(incomingVal, currForOp, thisHistorySum);
+
+      // get the value value assigned to the argument coming from the previous
+      // iteration
+      auto yieldOp = block->getTerminator();
+      auto prevVal = yieldOp->getOperand(arg.getArgNumber() - 1);
+      int min2 = minOverHistories(prevVal, yieldOp, thisHistorySum);
+      return std::min(std::min(min1, min2), minWaitNumber);
+    }
+    assert(false && "Unexpected Op case in minOverHistories.");
+    return 0;
+  };
+
+  return minOverHistories(extractOp.getOperand(0), extractOp, 0);
+}
+
+/// Insert wait ops after the extract_slice ops.
+void mlir::triton::MatmulPipelineSchedule::insertWaits(RewriterBase &rewriter,
+                                                       scf::ForOp forOp) {
+  for (auto op : extractOps) {
+    auto extractSlice = cast<ttg::ExtractSliceOp>(op);
```

Comment:
Correct, this what I have mentioned in the description as "changes the codegen a little". On H100 the difference is negligible, but on A100 I can actually see a perf drop because of this change. I will try to push the waits down, before the first use of the data. It may (or may not) actually help with perf on A100, because the first load from the memory will have one less commit to wait for. If I'll still see perf problems, I'll go ahead and merge the waits together.

Reply:
> I will try to push the waits down, before the first use of the data
I'm not sure I understood this part. Shouldn't the wait always be before the extract_slice, i.e. you cannot extract_slice then wait_group ?

Reply:
Ah that's right! So as long as the wait is before the convert_layout it should be fine, right ?